toolchain/check/thunk.cpp

```diff
@@ -0,0 +1,421 @@
+  auto decl_id = AddPlaceholderInst(
+      context, SemIR::LocIdAndInst::UncheckedLoc(loc_id, function_decl));
+  auto generic_id = BuildGenericDecl(context, decl_id);
+
+  // Create the `Function` object.
+  auto& signature = context.functions().Get(signature_id);
+  auto& callee = context.functions().Get(callee_id);
+  function_decl.function_id = context.functions().Add(
+      SemIR::Function{{.name_id = signature.name_id,
+                       .parent_scope_id = callee.parent_scope_id,
+                       .generic_id = generic_id,
+                       .first_param_node_id = signature.first_param_node_id,
+                       .last_param_node_id = signature.last_param_node_id,
+                       .pattern_block_id = pattern_block_id,
+                       .implicit_param_patterns_id = implicit_param_patterns_id,
+                       .param_patterns_id = param_patterns_id,
+                       .is_extern = false,
+                       .extern_library_id = SemIR::LibraryNameId::None,
+                       .non_owning_decl_id = SemIR::InstId::None,
+                       .first_owning_decl_id = decl_id,
+                       .definition_id = decl_id},
+                      {.call_params_id = call_params_id,
+                       .return_slot_pattern_id = return_slot_pattern_id,
+                       .virtual_modifier = callee.virtual_modifier,
+                       .virtual_index = callee.virtual_index,
+                       .self_param_id = self_param_id}});
+  function_decl.type_id =
+      GetFunctionType(context, function_decl.function_id,
+                      context.scope_stack().PeekSpecificId());
+  ReplaceInstBeforeConstantUse(context, decl_id, function_decl);
+  return {function_decl.function_id, decl_id};
+}
+
+// Build an expression that names the value matched by a pattern.
+static auto BuildPatternRef(Context& context, SemIR::FunctionId function_id,
+                            SemIR::InstId pattern_id) -> SemIR::InstId {
+  auto pattern = context.insts().Get(pattern_id);
+
+  auto addr = pattern.TryAs<SemIR::AddrPattern>();
+  if (addr) {
+    pattern_id = addr->inner_id;
+    pattern = context.insts().Get(pattern_id);
+  }
+
+  auto pattern_ref_id = SemIR::InstId::None;
+  if (auto value_param = pattern.TryAs<SemIR::ValueParamPattern>()) {
+    // Build a reference to this parameter.
+    auto call_param_id = context.inst_blocks().Get(
+        context.functions()
+            .Get(function_id)
+            .call_params_id)[value_param->index.index];
+    // Use a pretty name for the `name_ref`. While it's suspicious to use a
+    // pretty name in the IR like this, the only reason we include a name at
+    // all here is to make the formatted SemIR more readable.
+    pattern_ref_id = AddInst<SemIR::NameRef>(
+        context, SemIR::LocId(pattern_id),
+        {.type_id = context.insts().Get(call_param_id).type_id(),
```

Comment:
Minor, but a lot of this file uses the other style:

```suggestion
    pattern_ref_id = AddInst(
        context, SemIR::LocId(pattern_id),
        SemIR::NameRef{.type_id = context.insts().Get(call_param_id).type_id(),
```

Maybe we should generally discuss consolidating on one or the other?

Reply:
I've changed the other calls to use a style more like this one instead :) Also added some extra checking to make sure we're creating instructions with locations of the right kind.